examples/02-importing-exporting-module-declarations.md

```diff
@@ -0,0 +1,29 @@
+# Importing and Exporting Module Declarations
+
+Module declarations can be exported, re-exported and imported from other modules:
+
+```js
+// file-a.js
+
+export module modY {
+    export const y = 1;
+
+    export module modX {
+        export const x = 2;
+    }
+}
+
+export { modX } from modY;
+```
+
+```js
+// file-b.js
+
+import { modX, modY } from "./file-a.js";
+import { x } from modX;
+import { y } from modY;
```

Comment:
this is interesting. you can import from a first-class reified value?

what happens if i have:
```js
const { modX } = await import('./file-a.js');
import { x } from modX;
```
?

Reply:
this is a link error. `import { x } from modX` run in the link stage.

Reply:
Right, but that seems very confusing - how do users know when a first-class value can be statically imported from or not?

Reply:
they must be statically declared as a ModuleDeclaration. maybe the type checker will help

Reply:
There is no type checker - not everyone uses TS and proposals can't assume there will be a linter, let alone a type checker.

Reply:
this is like how you know the imported path is valid, or if the dependency graph contains a Top-Level-Await.

Reply:
Especially around modules, there are already different things that users now know where linters and type checkers help, and the language assumes that they do "the right thing" even if they don't use a linter (otherwise, it throws an error):
- As @Jack-Works mentioned, they have to write a correct import path otherwise the module will throw when loaded
- Users have to know the name of exports of a module (you cannot `import { x }` from a module that only exports `y`)
- Users have to know the type of exports of a module: in `const { f } = await import("mod"); f()` they need to know that `f` is a function, even if they don't use a type checker

> how do users know when a first-class value can be statically imported from or not?

In general you cannot "import from values": linking happens before execution, so before that those values even exist. Users who use import declarations will have to replace their "I can import from files" knowledge with "I can import from files or module declarations".

Additionally, it's already known that imports are hoisted to the top of the file[^1], or at least that imports are "executed" before any code in the file. If you think about JS modules as two sections:
```js
// all the imports get moved to this section
import { x } from modX;

// all the runtime code gets moved to this section
const { modX } = await import('./file-a.js');
```
then it becomes more evident that you cannot import from runtime values, because they are not initialized yet when the imports run.

[^1]: The reality is not that they are hoisted, but that they run in a completely different phase (linking vs evaluation). However, I have always only seen is taught as "they are hoisted" - and it's a good enough approximation of what happens.